### ISSUE TITLE
Rule proposal: Explicit undefined type for Record members

### ISSUE THREAD

**Issue by WesleyYue** [typescript-eslint/typescript-eslint#6263]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

_I looked really hard and couldn't find anything that does the following. Maybe it does already exist, or there is a good reason not to do this. Please lmk_

Currently, this code can lead to unsafe assignments during runtime if an undefined key is accessed:
```
const apple = {
  existingKey: "test"
}

function t(apple: Record<string, string>) {  // Should have a warning that it should be Record<string, string | undefined> instead
  const { existingKey, notExistingKey } = apple;  // existingKey: string; notExistingKey: string
  console.log(typeof existingKey === string); // true
  console.log(typeof notExistingKey === string); // false
}
```

Using Record<string, string> defines that any subsequent access to keys of that Record does exist, and silently places the responsibility on the programmer to ensure that is true. In many cases, what the programmer actually meant is something along the lines of "if any keys are accessed, they will be of type string, but I don't know what keys will exist. They might be undefined."

A strict rule that checks that all uses of Record explicitly label the member types as potentially undefined could help make code less error prone at runtime.

### Fail Cases

```typescript
const apple = {
  existingKey: "test"
}

function t(apple: Record<string, string>) {  // Should have a warning that it should be Record<string, string | undefined> instead
  const { existingKey, notExistingKey } = apple;  // existingKey: string; notExistingKey: string
  console.log(typeof existingKey === string); // true
  console.log(typeof notExistingKey === string); // false
}
```

### Pass Cases

```typescript
const apple = {
  existingKey: "test"
}

function t(apple: Record<string, string | undefined>) {  // Programmer explicitly states potentially undefined type. Pass
  const { existingKey, notExistingKey } = apple;  // existingKey: string | undefined; notExistingKey: string | undefined
  console.log(typeof existingKey === string); // true
  console.log(typeof existingKey === string); // false
}
```

### Additional Info

This would also need to be extended for "index signature" style types.

**Comment by bradzacher** [comment]
TypeScript already has an option for this - [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess). [Your example in the playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBAhgBwQGwKYwLwwN4CgYyoAeAltCWAOYDSqAngFwwBEUq0zuAvrrgGYBXMMCglwMKAApEKVEwBKqUACcAJgB5oyipQA0MLToB8AShz4YoSLGyFS5KrTr6wIKAFF7ox-RhdM8EhoANy8BMRk3jT0FgD0sQQEAHoA-GEwrh5eOk5xCYmp3EA)

TS's option is safer than unioning in `undefined` because TS's version only adds `undefined` at read time, and disallows it at write time.
The manual solution is bad because it allows this:
```ts
function t(apple: Record<string, string | undefined>) {  
  apple['key'] = undefined; // BAD BAD BAD - should not be allowed
}
```

We won't be adding a rule that duplicates a TS compiler option, esp when it enforces an unsafe practice!

**Comment by WesleyYue** [comment]
Awesome, thanks for the pointer to noUncheckedIndexedAccess, exactly what I was looking for.
